lib/common/strings.c

```diff
@@ -148,6 +150,123 @@ crm_parse_int(const char *text, const char *default_text)
     return (int) result;
 }
 
+/*!
+ * \internal
+ * \brief Scan a double-precision floating-point value from a string
+ *
+ * \param[in]      text         The string to parse
+ * \param[out]     result       Parsed value on success, or -1 on error
+ * \param[in]      default_text Default string to parse if \p text is
+ *                              \c NULL
+ * \param[out]     end_text     If not \c NULL, where to store a pointer
+ *                              to the position immediately after the
+ *                              value
+ *
+ * \return Standard Pacemaker return code (\c pcmk_rc_ok on success,
+ *         \c pcmk_rc_no_output on \c NULL \p result,
+ *         \c pcmk_rc_no_input on no input string,
+ *         \c pcmk_rc_unknown_format on failed string conversion due to
+ *         invalid input, \c pcmk_rc_after_range on arithmetic overflow,
+ *         \c pcmk_rc_before_range on arithmetic underflow, or \c errno
+ *         from \c strtod() on other parse errors)
+ */
+int
+pcmk__scan_double(const char *text, double *result, const char *default_text,
+                  char **end_text)
+{
+    int rc = pcmk_rc_ok;
+    char *local_end_text = NULL;
+
+    CRM_CHECK(result != NULL, return pcmk_rc_no_output);
```

Comment:
I guessed on some of these return codes. Let me know if any of them don't make sense or if there are better choices. I'm returning all pacemaker return codes now instead of returning errno's.

Reply:
The standard codes are relatively new, so there is some guesswork involved. I usually scan errno(3) first then do "ack --hh pcmk_rc_" to see existing choices.

In this particular case, EINVAL is the go-to choice for incorrectly supplied arguments. There's nothing wrong with repurposing an existing code if its name sounds reasonably appropriate, but so far pcmk_rc_no_output's use has been related to output in the sense of printing output to the user. Adding new codes is encouraged when a new message would make the user's life easier.

Reply:
I realized after I called it a night that there were probably string descriptions of the pcmk return codes. And indeed there are. Those are helpful, and would have steered me away from pcmk_rc_no_output if I had looked at them :)

Reply:
Are pcmk_rc_{after,before}_range reasonable for differentiating overflow and underflow respectively? Or do you think ERANGE makes more sense for both (as strtod() does)?

Reply:
Good question, by name either could make sense. However these are the POSIX errnos with their system messages:

* ERANGE: Result too large
* EOVERFLOW: Value too large to be stored in data type

It's interesting that there are two standard errno's for "too large" and none for "too small". There doesn't seem to be a consistently observed difference between the above two, but a decent interpretation is that ERANGE would be for a logically invalid value (e.g. 13 for month), whereas EOVERFLOW is for a logically valid value that can't be represented.

strtod() is an interesting case. It's not defined by POSIX but by C99, which specifies ERANGE as its errno, but perror() on a POSIX system (or equivalently pcmk_rc_str() in pacemaker) would certainly be confusing for underflow.

The pacemaker range codes are similar to the conception of ERANGE as logically invalid, but distinguishing "too small" and "too large", and not necessarily being invalid but just occurring outside the range (e.g. January 5 is not in March through May).

I would say EOVERFLOW is an obvious choice for overflow, but what for underflow? I would say a new pcmk_rc_underflow, or just (ab)use EOVERFLOW and tell callers not to use pcmk_rc_str() if *result is 0.

Reply:
That's enlightening. A quick search for EOVERFLOW vs. ERANGE shows this has indeed been ambiguous for a while. ERANGE seems to have been historically used a lot of places where EOVERFLOW is a better fit, due to the way ERANGE is used in the C standard.

> I would say a new pcmk_rc_underflow, or just (ab)use EOVERFLOW and tell callers not to use pcmk_rc_str() if *result is 0.

It's no trouble to add a pcmk_rc_underflow, so that seems like a better route than abusing EOVERFLOW. (My need for symmetry is dying to add a pcmk_rc_overflow, but then we would be triply redundant ;) ).

Reply:
Done and pushed.